src/modules/mandatory-webhook/mandatory-webhook.service.ts

```diff
@@ -0,0 +1,181 @@
+      );
+    }
+
+    return valid;
+  }
+
+  public handleCustomersDataRequestWebhook(
+    req: RawBodyRequest<Request>,
+  ): Promise<void> {
+    return this.handleWebhookRequest(
+      req,
+      this.emailService.sendCustomerDataRequest,
+      'Webhook for requesting customers data from the shop with id',
+      'An error occurs during requesting customers data from the shop with id',
+    );
+  }
+
+  public handleCustomersRedactRequestWebhook(
+    req: RawBodyRequest<Request>,
+  ): Promise<void> {
+    return this.handleWebhookRequest(
+      req,
+      this.emailService.sendCustomerDataRequest,
+      'Webhook for requesting customers redact from the shop with id',
+      'An error occurs during requesting customers redact from the shop with id',
+    );
+  }
+
+  public async handleUninstallAppWebhook(
+    req: RawBodyRequest<Request>,
+  ): Promise<void> {
+    const webhookId = req.headers['x-shopify-webhook-id'] as string;
+
+    if (await this.webhookService.isDuplicate(webhookId)) return;
+
+    const { id: shopId } = req.body;
+
+    Logger.debug(
+      `Webhook for uninstalling app from the shop with id: ${shopId}. ${JSON.stringify(
+        {
+          body: req.body,
+          headers: req.headers,
+        },
+      )}`,
+    );
+
+    try {
+      await this.uninstallApp(shopId);
+
+      Logger.debug(
+        `App was successfully uninstalled from the shop with id: ${shopId}`,
+      );
+
+      await this.saveWebhook(req);
+    } catch (error) {
+      Logger.debug(
+        `An error occurs during uninstalling app from the shop with id: ${shopId}: ${JSON.stringify(
+          {
+            error,
+          },
+          null,
+          2,
+        )}`,
+      );
+    }
+  }
+
+  private async handleWebhookRequest(
+    req: RawBodyRequest<Request>,
+    emailServiceMethod: (data: string) => Promise<void>,
+    loggerInfo: string,
+    loggerError: string,
+  ): Promise<void> {
+    const webhookId = req.headers['x-shopify-webhook-id'] as string;
+    const data = JSON.stringify(
+      {
+        body: req.body,
+        headers: req.headers,
+      },
+      null,
+      2,
+    );
+
+    if (await this.webhookService.isDuplicate(webhookId)) return;
+
+    const { id: shopId } = req.body;
+
+    Logger.debug(
+      `${loggerInfo}: ${shopId}. ${JSON.stringify({
+        body: req.body,
+        headers: req.headers,
+      })}`,
+    );
+
+    try {
+      await this.saveWebhook(req);
+      await emailServiceMethod(data);
+    } catch (error) {
+      Logger.debug(
+        `${loggerError}: ${shopId}: ${JSON.stringify(
+          {
+            error,
+          },
+          null,
+          2,
+        )}`,
+      );
+    }
+  }
+
+  private saveWebhook(req: RawBodyRequest<Request>): Promise<Webhook> {
+    const expiredAt = new Date();
+
+    // Add 6 months
+    expiredAt.setMonth(expiredAt.getMonth() + 6);
```

Comment:
with Postgres we don't have TTL

Reply:
I know but if we decide to get all hooks that are already expired and work with them then I suppose this solution is ok. Or should I delete this field?

Reply:
remove it